File: apple/Chuchotage/Network/RealtimeTranslationClientSecretProvider.swift
import Foundation

final class RealtimeTranslationClientSecretProvider: @unchecked Sendable {
    private let urlSession: URLSession
    private let clientSecretURL: URL
    private let retryDelaysNanoseconds: [UInt64]
    private let sponsoredTrialClient: SponsoredTrialClient

    init(
        urlSession: URLSession = .shared,
        clientSecretURL: URL = URL(string: "https://api.openai.com/v1/realtime/translations/client_secrets")!,
        sponsoredTrialClient: SponsoredTrialClient = SponsoredTrialClient(),
        retryDelaysNanoseconds: [UInt64] = [
            250_000_000,
            500_000_000,
            1_000_000_000,
            2_000_000_000,
        ]
    ) {
        self.urlSession = urlSession
        self.clientSecretURL = clientSecretURL
        self.sponsoredTrialClient = sponsoredTrialClient
        self.retryDelaysNanoseconds = retryDelaysNanoseconds
    }

    func sessionBearerToken(
        for credential: OpenAICredential,
        targetLanguageCode: String = TranslationLanguages.defaultTargetLanguageCode,
        refreshCredentialAfterUnauthorized: @Sendable () async throws -> OpenAICredential? = { nil }
    ) async throws -> RealtimeTranslationSessionToken {
        let sanitizedLanguage = TranslationLanguages.sanitizeOutputLanguageCode(targetLanguageCode)

        switch credential.kind {
        case .apiKey:
            return RealtimeTranslationSessionToken(
                value: credential.value,
                shouldSendSessionUpdate: true
            )

        case .chatGPTAccessToken:
            let clientSecret = try await createClientSecretWithRefreshRetry(
                credential: credential,
                targetLanguageCode: sanitizedLanguage,
                refreshCredentialAfterUnauthorized: refreshCredentialAfterUnauthorized
            )
            return RealtimeTranslationSessionToken(
                value: clientSecret,
                shouldSendSessionUpdate: false
            )

        case .sponsoredTrial:
            do {
                return try await sponsoredTrialClient.sessionBearerTokenFor(
                    installID: credential.value,
                    targetLanguageCode: sanitizedLanguage,
                    // Keep original-speech transcript parity with iOS transcript panes.
                    sourceTranscriptEnabled: true
                )
            } catch let error as SponsoredTrialClientError {
                switch error {
                case .requestFailed(let message):
                    throw RealtimeTranslationClientSecretError.requestFailed(message)
                case .network(let message):
                    throw RealtimeTranslationClientSecretError.network(message)
                }
            } catch {
                throw RealtimeTranslationClientSecretError.requestFailed(
                    "Sponsored trial request failed."
                )
            }
        }
    }

    private func createClientSecretWithRefreshRetry(
        credential: OpenAICredential,
        targetLanguageCode: String,
        refreshCredentialAfterUnauthorized: @Sendable () async throws -> OpenAICredential?
    ) async throws -> String {
        do {
            return try await createClientSecret(
                accessToken: credential.value,
                targetLanguageCode: targetLanguageCode
            )
        } catch RealtimeTranslationClientSecretError.unauthorized {
            let refreshed = try await refreshCredentialAfterUnauthorized()
            guard
                let refreshed,
                refreshed.kind == .chatGPTAccessToken,
                !OpenAICredentialValidator.normalize(refreshed.value).isEmpty
            else {
                throw RealtimeTranslationClientSecretError.unauthorized
            }

            return try await createClientSecret(
                accessToken: OpenAICredentialValidator.normalize(refreshed.value),
                targetLanguageCode: targetLanguageCode
            )
        }
    }

    private func createClientSecret(
        accessToken: String,
        targetLanguageCode: String
    ) async throws -> String {
        var request = URLRequest(url: clientSecretURL)
        request.httpMethod = "POST"
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue(OpenAIRequestHeaders.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try RealtimeTranslationRequestBuilder.clientSecretRequestBody(
            targetLanguageCode: targetLanguageCode
        )

        return try await executeClientSecretRequestWithRetry(request)
    }

    private func executeClientSecretRequestWithRetry(_ request: URLRequest) async throws -> String {
        for attempt in retryDelaysNanoseconds.indices {
            do {
                let (data, response) = try await urlSession.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw RealtimeTranslationClientSecretError.requestFailed(
                        "OpenAI returned an invalid response."
                    )
                }

                let text = String(data: data, encoding: .utf8) ?? ""
                let payload = parseJsonObject(data)

                switch httpResponse.statusCode {
                case 200..<300:
                    if let value = payload?["value"] as? String, !value.isEmpty {
                        return value
                    }
                    throw RealtimeTranslationClientSecretError.requestFailed(
                        "OpenAI did not return a translation client secret."
                    )

                case 401:
                    throw RealtimeTranslationClientSecretError.unauthorized

                case 429, 500...599:
                    if attempt < retryDelaysNanoseconds.index(before: retryDelaysNanoseconds.endIndex) {
                        try await Task.sleep(nanoseconds: retryDelaysNanoseconds[attempt])
                        continue
                    }
                    throw RealtimeTranslationClientSecretError.requestFailed(
                        "Could not create a translation client secret. Check the network and try again."
                    )

                default:
                    throw RealtimeTranslationClientSecretError.requestFailed(
                        errorMessage(from: payload, fallbackText: text)
                    )
                }
            } catch let error as RealtimeTranslationClientSecretError {
                throw error
            } catch let error as URLError {
                if isRetryableNetworkError(error),
                   attempt < retryDelaysNanoseconds.index(before: retryDelaysNanoseconds.endIndex) {
                    try await Task.sleep(nanoseconds: retryDelaysNanoseconds[attempt])
                    continue
                }
                throw RealtimeTranslationClientSecretError.network(networkMessage(for: error))
            }
        }

        throw RealtimeTranslationClientSecretError.requestFailed(
            "Failed to create translation client secret."
        )
    }

    private func parseJsonObject(_ data: Data) -> [String: Any]? {
        try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private func errorMessage(from payload: [String: Any]?, fallbackText: String) -> String {
        if
            let error = payload?["error"] as? [String: Any],
            let message = error["message"] as? String,
            !message.isEmpty
        {
            return message
        }

        if let message = payload?["error"] as? String, !message.isEmpty {
            return message
        }

        return fallbackText.isEmpty ? "Failed to create translation client secret." : fallbackText
    }

    private func isRetryableNetworkError(_ error: URLError) -> Bool {
        switch error.code {
        case .cannotFindHost,
             .dnsLookupFailed,
             .timedOut,
             .cannotConnectToHost,
             .networkConnectionLost,
             .notConnectedToInternet:
            return true
        default:
            return false
        }
    }

    private func networkMessage(for error: URLError) -> String {
        switch error.code {
        case .cannotFindHost, .dnsLookupFailed:
            return L10n.string(
                "network.resolveOpenAI",
                defaultValue: "Could not resolve api.openai.com. Check VPN, DNS, or network settings and try again."
            )
        default:
            return L10n.string(
                "network.reachOpenAI",
                defaultValue: "Could not reach api.openai.com. Check the network and try again."
            )
        }
    }
}

enum RealtimeTranslationClientSecretError: LocalizedError, Sendable {
    case unauthorized
    case requestFailed(String)
    case network(String)

    var errorDescription: String? {
        switch self {
        case .unauthorized:
            return L10n.string("chatGPTSignIn.expired", defaultValue: "ChatGPT sign-in expired. Sign in again.")
        case .requestFailed(let message), .network(let message):
            return message
        }
    }
}

File: app/src/main/java/com/andreabertoncini/chuchotage/network/RealtimeTranslationClientSecretProvider.kt
package com.andreabertoncini.chuchotage.network

import com.andreabertoncini.chuchotage.settings.TranslationLanguages
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

data class RealtimeTranslationSessionToken(
    val value: String,
    val shouldSendSessionUpdate: Boolean,
)

class RealtimeTranslationClientSecretProvider(
    private val okHttpClient: OkHttpClient = OkHttpClient(),
    private val clientSecretUrl: String = CLIENT_SECRET_URL,
    private val retryDelaysMs: LongArray = CLIENT_SECRET_RETRY_DELAYS_MS,
) {
    suspend fun sessionBearerTokenFor(
        credential: OpenAiCredential,
        targetLanguageCode: String = TranslationLanguages.DEFAULT_TARGET_LANGUAGE_CODE,
        sourceTranscriptEnabled: Boolean = false,
        refreshCredentialAfterUnauthorized: suspend () -> OpenAiCredential? = { null },
    ): RealtimeTranslationSessionToken {
        val sanitizedTargetLanguageCode = TranslationLanguages.sanitizeOutputLanguageCode(targetLanguageCode)
        return when (credential.kind) {
            OpenAiCredentialKind.API_KEY -> RealtimeTranslationSessionToken(
                value = credential.value,
                shouldSendSessionUpdate = true,
            )
            OpenAiCredentialKind.CHATGPT_ACCESS_TOKEN -> {
                val clientSecret = createClientSecretWithRefreshRetry(
                    credential = credential,
                    targetLanguageCode = sanitizedTargetLanguageCode,
                    sourceTranscriptEnabled = sourceTranscriptEnabled,
                    refreshCredentialAfterUnauthorized = refreshCredentialAfterUnauthorized,
                )
                RealtimeTranslationSessionToken(
                    value = clientSecret,
                    shouldSendSessionUpdate = false,
                )
            }
            OpenAiCredentialKind.SPONSORED_TRIAL -> error(
                "Sponsored trial client secrets are requested from the Chuchotage backend.",
            )
        }
    }

    private suspend fun createClientSecretWithRefreshRetry(
        credential: OpenAiCredential,
        targetLanguageCode: String,
        sourceTranscriptEnabled: Boolean,
        refreshCredentialAfterUnauthorized: suspend () -> OpenAiCredential?,
    ): String {
        return try {
            createClientSecret(credential.value, targetLanguageCode, sourceTranscriptEnabled)
        } catch (exception: ClientSecretUnauthorizedException) {
            val refreshed = try {
                refreshCredentialAfterUnauthorized()
            } catch (refreshFailure: ChatGptSignInFailure.ReauthenticationRequired) {
                throw ClientSecretUnauthorizedException(refreshFailure.message ?: REAUTHENTICATE_MESSAGE, refreshFailure)
            }

            if (
                refreshed?.kind == OpenAiCredentialKind.CHATGPT_ACCESS_TOKEN &&
                refreshed.value.isNotBlank() &&
                refreshed.value != credential.value
            ) {
                createClientSecret(refreshed.value, targetLanguageCode, sourceTranscriptEnabled)
            } else {
                throw ClientSecretUnauthorizedException(REAUTHENTICATE_MESSAGE, exception)
            }
        }
    }

    private suspend fun createClientSecret(
        accessToken: String,
        targetLanguageCode: String,
        sourceTranscriptEnabled: Boolean,
    ): String = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(clientSecretUrl)
            .addHeader("Authorization", "Bearer $accessToken")
            .addHeader("User-Agent", OpenAiRequestHeaders.userAgent)
            .post(
                clientSecretRequestBody(
                    targetLanguageCode = targetLanguageCode,
                    sourceTranscriptEnabled = sourceTranscriptEnabled,
                ).toString().toRequestBody(JSON_MEDIA_TYPE),
            )
            .build()

        executeClientSecretRequestWithRetry(request)
    }

    private suspend fun executeClientSecretRequestWithRetry(request: Request): String {
        var lastNetworkError: IOException? = null

        retryDelaysMs.forEachIndexed { attempt, delayMs ->
            try {
                okHttpClient.newCall(request).execute().use { response ->
                    val text = response.body?.string().orEmpty()
                    val payload = parseJsonObject(text)

                    when {
                        response.isSuccessful -> {
                            return payload
                                ?.optString("value")
                                ?.takeIf { it.isNotBlank() }
                                ?: error("OpenAI did not return a translation client secret.")
                        }
                        response.code == 401 -> throw ClientSecretUnauthorizedException(REAUTHENTICATE_MESSAGE)
                        response.code == 429 || response.code in 500..599 -> {
                            if (attempt < retryDelaysMs.lastIndex) {
                                delay(delayMs)
                            } else {
                                throw ClientSecretRequestException(
                                    "Could not create a translation client secret. Check the phone's network and try again.",
                                )
                            }
                        }
                        else -> throw ClientSecretRequestException(errorMessageFrom(payload, text))
                    }
                }
            } catch (exception: IOException) {
                lastNetworkError = exception
                if (!exception.isRetryableClientSecretNetworkError() || attempt == retryDelaysMs.lastIndex) {
                    throw ClientSecretRequestException(exception.toClientSecretNetworkMessage(), exception)
                }
                delay(delayMs)
            }
        }

        val exception = lastNetworkError ?: IOException("Network request failed.")
        throw ClientSecretRequestException(exception.toClientSecretNetworkMessage(), exception)
    }

    internal fun clientSecretRequestBody(
        targetLanguageCode: String,
        sourceTranscriptEnabled: Boolean = false,
    ): JSONObject {
        return JSONObject()
            .put(
                "expires_after",
                JSONObject()
                    .put("anchor", "created_at")
                    .put("seconds", CLIENT_SECRET_TTL_SECONDS),
            )
            .put(
                "session",
                JSONObject()
                    .put("model", "gpt-realtime-translate")
                    .put(
                        "audio",
                        JSONObject()
                            .put(
                                "input",
                                inputAudioConfig(sourceTranscriptEnabled),
                            )
                            .put(
                                "output",
                                JSONObject().put(
                                    "language",
                                    TranslationLanguages.sanitizeOutputLanguageCode(targetLanguageCode),
                                ),
                            ),
                    ),
            )
    }

    private fun inputAudioConfig(sourceTranscriptEnabled: Boolean): JSONObject {
        return JSONObject()
            .apply {
                if (sourceTranscriptEnabled) {
                    put(
                        "transcription",
                        JSONObject().put("model", "gpt-realtime-whisper"),
                    )
                }
            }
            .put("noise_reduction", JSONObject.NULL)
    }

    private fun parseJsonObject(text: String): JSONObject? {
        return runCatching { JSONObject(text) }.getOrNull()
    }

    private fun errorMessageFrom(payload: JSONObject?, text: String): String {
        return payload
            ?.optJSONObject("error")
            ?.optString("message")
            ?.takeIf { it.isNotBlank() }
            ?: payload?.optString("error")?.takeIf { it.isNotBlank() }
            ?: text.takeIf { it.isNotBlank() }
            ?: "Failed to create translation client secret."
    }

    private fun IOException.isRetryableClientSecretNetworkError(): Boolean {
        return this is UnknownHostException || this is SocketTimeoutException
    }

    private fun IOException.toClientSecretNetworkMessage(): String {
        return if (this is UnknownHostException) {
            "Could not resolve api.openai.com. Check the phone's VPN, private DNS, or network and try again."
        } else {
            "Could not reach api.openai.com. Check the phone's network and try again."
        }
    }

    private class ClientSecretUnauthorizedException(
        message: String,
        cause: Throwable? = null,
    ) : IllegalStateException(message, cause)

    private class ClientSecretRequestException(
        message: String,
        cause: Throwable? = null,
    ) : IllegalStateException(message, cause)

    companion object {
        private const val CLIENT_SECRET_URL = "https://api.openai.com/v1/realtime/translations/client_secrets"
        private const val CLIENT_SECRET_TTL_SECONDS = 600
        private const val REAUTHENTICATE_MESSAGE = "ChatGPT sign-in expired. Sign in again."
        private val CLIENT_SECRET_RETRY_DELAYS_MS = longArrayOf(250, 500, 1_000, 2_000)
        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }
}
